pjsip/src/pjsua2/endpoint.cpp

```diff
@@ -2493,3 +2579,63 @@ void Endpoint::handleIpChange(const IpChangeParam &param) PJSUA2_THROW(Error)
     pjsua_ip_change_param ip_change_param = param.toPj();
     PJSUA2_CHECK_EXPR(pjsua_handle_ip_change(&ip_change_param));
 }
+
+#if PJSIP_HAS_DIGEST_AKA_AUTH
+pj_status_t  Endpoint::createAkaResponse(OnCredAuthParam &param) PJSUA2_THROW(Error)
+{
+    pj_caching_pool cp;
+    pj_pool_t *pool;
+    pj_caching_pool_init(&cp, NULL, 0);
+    pool = pj_pool_create(&cp.factory, "onCredAuth", 1024, 1024, NULL);
+    if (!pool)
+    return (PJ_ENOMEM);
+
+    pjsip_digest_challenge chal = param.digestChallenge.toPj();
+    pjsip_cred_info cred = param.credentialInfo.toPj();
+    const pj_str_t method = str2Pj(param.method);
+    pjsip_digest_credential auth = param.digestCredential.toPj();
+
+    pj_status_t status = pjsip_auth_create_aka_response(pool, &chal, &cred, &method, &auth);
+
+    DigestCredential digestCred;
+    digestCred.fromPj(auth);
+    // Copy so that we are able to release the pool.
+    param.digestCredential = digestCred;
+    pj_pool_release(pool);
+
+    return status;
+}
+#endif
+
+pj_status_t Endpoint::pjsip_auth_create_aka_response_callback(pj_pool_t *pool,
+                                           const pjsip_digest_challenge *chal,
+                                           const pjsip_cred_info *cred,
+                                           const pj_str_t *method,
+                                           pjsip_digest_credential *auth)
+{
+    OnCredAuthParam prm;
+    prm.digestChallenge.fromPj(*chal);
+    prm.credentialInfo.fromPj(*cred);
+    prm.method = pj2Str(*method);
+    prm.digestCredential.fromPj(*auth);
+
+    pj_status_t status = Endpoint::instance().onCredAuth(prm);
+
+   if (status == PJ_SUCCESS) {
+	    pjsip_digest_credential auth_new = prm.digestCredential.toPj();
+	    // Duplicate in the pool, so that digestCredential
+	    // is allowed to be destructed at the end of the method.
+	    pj_strdup(pool, &auth->realm, &auth_new.realm);
+	    pj_strdup(pool, &auth->username, &auth_new.username);
+	    pj_strdup(pool, &auth->nonce, &auth_new.nonce);
+	    pj_strdup(pool, &auth->uri, &auth_new.uri);
+	    pj_strdup(pool, &auth->response, &auth_new.response);
+	    pj_strdup(pool, &auth->algorithm, &auth_new.algorithm);
+	    pj_strdup(pool, &auth->cnonce, &auth_new.cnonce);
+	    pj_strdup(pool, &auth->opaque, &auth_new.opaque);
+	    pj_strdup(pool, &auth->qop, &auth_new.qop);
+	    pj_strdup(pool, &auth->nc, &auth_new.nc);
+	    pjsip_param_clone(pool, &auth->other_param, &auth_new.other_param);
+    }
+    return status;
+}
```

Comment:
For simplicity (specifical for PJSUA2), what if we directly call `pjsip_auth_create_aka_response()` from this PJSIP callback when `onCredAuth()` returning `PJ_ENOTSUP` (as in base implementation), so we can omit `createAkaResponse()`. Need to specify this behavior in the `onCredAuth()` docs/spec though.

Reply:
So opt-out instead of opt-in. Sounds good.